Add qt6 support

biagas opened the pull request:
### Description

Add bv_qt6 module.  It only downloads and compiles the modules needed by VisIt (base, tools and svg).
Qt 6 can be requested by '--qt6'.  If vtk is also being built, version 9 will be forced.

Updated the FindVisItQt module to support Qt 6.

Fixed code within VisIt to do a Qt version check for code removed or deprecated in Qt6 but not deprecated in previous versions.

Since newer VS compiler was needed to build Qt on Windows, all TP libraries were rebuilt with MSVC2022, and Windows config-site updated to support.  Only Qt 6 and VTK 9 are available with MSVC2022.

Added Qt 6 support to toss 4 systems on CZ, and updated poodle18 config-site. To use QT 6, set  `USE_QT6` CMake variable to true for your build.  This is still considered experimental since it requires VTK 9 and we are still working out some issues with VTK 9 and Mesa.

### Type of change

<!-- Please check one of the boxes below -->

~~* [ ] Bug fix~~
* [X] New feature~~
~~* [ ] Documentation update~~
~~* [ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?

Compile cleanly on pascal with 'USE_QT6' unset (meaning qt5 is used), and with it set to true.
Compile cleanly on Windows with both MSVC2017 (qt5) and MSC2022 (qt6).

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [X] I have commented my code where applicable.~~
- [X] I have updated the release notes.~~
~~- [ ] I have made corresponding changes to the documentation.~~
~~- [ ] I have added debugging support to my changes.~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works.~~
- [X] I have confirmed new and existing unit tests pass locally with my changes.~~
~~- [ ] I have added new baselines for any new tests to the repo.~~
~~- [ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

biagas committed:
Add support for Qt 6.

biagas committed:
Add support for Qt 6.

biagas committed:
Misc Qt6 fixes.

biagas committed:
Misc Qt6 fixes.

biagas committed:
Add qt6 support to build_visit.
Force VTK9 when building QT6.
Update qwt version for qt6.

biagas committed:
Add qt6 support to build_visit.
Force VTK9 when building QT6.
Update qwt version for qt6.

biagas committed:
Change when config-site is included for Windows.

Move it to after the 'project' call so that compiler information is known.
This allows support of multiple versions of MSVC for pre-built TP libs, where perhaps the TP versions are different depending on compiler.
Made the inclusion of the config-site a macro to reduce code duplication.

biagas committed:
Change when config-site is included for Windows.

Move it to after the 'project' call so that compiler information is known.
This allows support of multiple versions of MSVC for pre-built TP libs, where perhaps the TP versions are different depending on compiler.
Made the inclusion of the config-site a macro to reduce code duplication.

biagas committed:
Change FindTiff.cmake so FindVTK9 doesn't complain.

biagas committed:
Change FindTiff.cmake so FindVTK9 doesn't complain.

biagas committed:
Add MSVC2022 support to windows-config.

In support of Qt 6 port.

biagas committed:
Add MSVC2022 support to windows-config.

In support of Qt 6 port.

biagas committed:
Merge branch 'task/biagas/add_qt6_support' of https://github.com/visit-dav/visit into task/biagas/add_qt6_support

biagas committed:
Merge branch 'task/biagas/add_qt6_support' of https://github.com/visit-dav/visit into task/biagas/add_qt6_support

biagas committed:
Fix CMAKE location for bv_qt6.
Ensures Qt6 can build where system cmake doesn't exist.

biagas committed:
More build_visit updates for QT6

set DO_QT to "no" if DO_QT6 is enabled.
Fix libvtklibxml2 location (lib64 vs lib) when building VTK9.

biagas committed:
Update release notes with Qt 6 support.
Remove notation from 3.3.4 and put in 3.4.0.

biagas committed:
Merge branch 'task/biagas/add_qt6_support' of https://github.com/visit-dav/visit into task/biagas/add_qt6_support

biagas committed:
Merge branch 'develop' into task/biagas/add_qt6_support

biagas committed:
Use idClicked instead of buttonClicked for QButtonGroup with Qt6.

biagas committed:
Merge branch 'develop' into task/biagas/add_qt6_support

biagas committed:
Update FindQt5.cmake

Put in missing OPENGL_LIBRARIES for QT_QTGUI_LIBRARY.

biagas committed:
Add Qt 6 support to poodle18.cmake.
Enabled by setting CMake var USE_QT6 to ON.
Fix comments and opengl issues with FindQt modules.

biagas committed:
Merge branch 'task/biagas/add_qt6_support' of https://github.com/visit-dav/visit into task/biagas/add_qt6_support

biagas committed:
Reconsolidate QT find modules into one supporting both version 5 and 6.

biagas committed:
Merge branch 'develop' into task/biagas/add_qt6_support

JustinPrivitera commented on src/CMake/FindSilo.cmake:
```diff
@@ -44,7 +44,11 @@
 #
 
 IF (WIN32)
-  SET_UP_THIRD_PARTY(SILO LIBS silohdf5)
+  if(EXISTS ${SILO_DIR}/lib/silohdf5.lib)
+      SET_UP_THIRD_PARTY(SILO LIBS silohdf5)
+  else() 
+      SET_UP_THIRD_PARTY(SILO LIBS siloh5)
+  endif()
```
modification comment?

biagas replied:
I had to rebuild silo (and all TP libs) with the newer compiler on Windows as part of this effort. I used a newer silo version which uses a different naming convention for the library.
Added a comment.

JustinPrivitera commented on src/winutil/QvisColorTableButton.C:
```diff
@@ -164,7 +164,7 @@ QvisColorTableButton::~QvisColorTableButton()
 QSize
 QvisColorTableButton::sizeHint() const
 {
-     return QSize(125, 40).expandedTo(QApplication::globalStrut());
+     return QSize(125, 40);
```
modification comment for this change?

biagas replied:
Added.

JustinPrivitera commented on src/winutil/QvisNoDefaultColorTableButton.C:
```diff
@@ -164,7 +164,7 @@ QvisNoDefaultColorTableButton::~QvisNoDefaultColorTableButton()
 QSize
 QvisNoDefaultColorTableButton::sizeHint() const
 {
-     return QSize(125, 40).expandedTo(QApplication::globalStrut());
+     return QSize(125, 40);
```
modification comment for this change?

biagas replied:
Added

JustinPrivitera commented on src/tools/dev/xmledit/XMLDocument.C:
```diff
@@ -140,17 +111,7 @@ XMLDocument::open(const QString &file)
-        bool success = reader.parse(source);
-
-        if (!success)
+        if (!parser.parse())
```
modification comment for these removals?

biagas replied:
Added.

JustinPrivitera commented on src/gui/QvisMainWindow.C:
```diff
@@ -428,12 +431,12 @@ QvisMainWindow::QvisMainWindow(int orientation, const char *captionString)
     {
         selFileAct = filePopup->addAction(*openIcon, tr("Select &file . . ."),
                          this, SIGNAL(activateFileWindow()),
-                         QKeySequence(Qt::CTRL + Qt::Key_F));
+                         QKeySequence(Qt::CTRL | Qt::Key_F));
```
Just curious, what does this do? Does this change functionality so that instead of pressing `ctrl` + `F` you can press either? Why make this change?

biagas replied:
Qt changed from using '+' to '|' for the same functionality.

JustinPrivitera commented on src/operators/LineSampler/QvisLineSamplerWindow.C:
```diff
@@ -99,7 +101,7 @@ QvisLineSamplerWindow::CreateWindowContents()
     int width = 100;
 
     propertyTabs = new QTabWidget(central);
-    topLayout->addWidget(propertyTabs, 1, 0);
+    topLayout->addWidget(propertyTabs, 1);
```
Why remove this argument? Is the last 0 for a Qt::Alignment?

biagas replied:
The addWidget method will no longer take '0' for the last argument. It has a default which is equivalent to what passing 0 used to mean.

JustinPrivitera commented on src/resources/help/en_US/relnotes3.4.0.html:
```diff
@@ -120,6 +120,9 @@
   <li>Support for boolArray and boolVector was added to xml code gen tools.</li>
   <li>Updated the "Creating a Release" section of the developer manual to describe the current process.</li>
+  <li>Add support for Qt6.  The build_visit option is '--qt6', which will build qt version 6.4.2, and also force building of VTK-9 if '--vtk' is also enabled.</li>
+  <li>Developer Changes 2</li>
```
?

biagas replied:
Was I not clear enough in the comment? Or is your '?' something different? Can you clarify?

JustinPrivitera replied:
Sorry - I am confused why you added "Developer Changes 2".

biagas replied:
Gotcha. Will fix that.

JustinPrivitera commented on src/tools/dev/clipeditor/ClipEditor.C:
```diff
@@ -702,8 +702,7 @@ ClipEditor::keyPressEvent(QKeyEvent *kev)
     }
-
-    viewer->updateGL();
+    viewer->update();
```
What does this change do? Modification comment?

JustinPrivitera reviewed (dismissed):
Just a few comments and questions.

I feel for you and the many times you had to replace buttonClicked with idClicked :(

biagas committed:
Add comments per reviewer request.

biagas committed:
Merge branch 'develop' into task/biagas/add_qt6_support

biagas committed:
Update relnotes3.4.0.html

Remove extraneous line.

JustinPrivitera commented:
I checked out this branch and I ran build visit with this invocation:
```
./build_visit --makeflags -j24 --mesagl --mpich --qt6 --vtk --required --optional --no-gdal
```

I saw this worrying message from the terminal partway through building the TPLs:
```
Building Qt (~60 minutes)
Unzipping/Untarring qt-everywhere-src-5.14.2.tar.xz . . .
Patching qt . . .
Patching qt 5.14.2 for numeric-limits
Patching qt 5.14.2 for qjp2handler cmath include
Patching qt 5.14.2 for Linux and Mesa-as-GL
Configuring Qt5: 
Building Qt5 . . . (~60 minutes)
Installing Qt5 . . . 
Done with Qt5
Done building Qt
```
Shouldn't we not be building qt5? Is it building qt5 because I turned on `--optional`? If qt6 is turned on should we disable the qt5 option?

Then this happens:
```
library qwt depends on qt6
library qt6 depends on mesagl glu
mesagl is already installed, skipping
glu is already installed, skipping
Building QT6 base (~10 minutes)
Unzipping/Untarring qtbase-everywhere-src-6.4.2.tar.xz . . .
Configuring Qt6 base: . . . 
Building Qt6 base . . . 
Qt6 base build failed. Giving up
Unable to build or install QT6. Bailing out.
Error in build process. See build_visit_log for more information. If the error is unclear, please include build_visit_log and contact the VisIt project via https://visit-help.llnl.gov. You may need to compress the build_visit_log using a program like gzip so it will fit within the size limits for attachments.
Log file full path:  /usr/workspace/justin/visit_builds/qt6-develop/build_visit_log
```
This is what the log says:
```
CMake Error at CMakeLists.txt:40 (message):
  The path
  "/usr/workspace/justin/visit_builds/qt6-develop/qtbase-everywhere-src-6.4.2-build"
  contains symlinks.  This is not supported.  Possible solutions:

              - map directories using a transparent mechanism such as mount --bind
              - pass the real path of the build directory to CMake, e.g. using             cd $(realpath <path>) before invoking cmake <source_dir>.
Call Stack (most recent call first):
  CMakeLists.txt:47 (qt_internal_check_if_path_has_symlinks)


'/usr/WS2/justin/visit_builds/qt6-develop/third_party/cmake/3.24.3/linux-x86_64_gcc-7.3/bin/cmake' '-DOPENGL_INCLUDE_DIR:PATH=/usr/workspace/justin/visit_builds/qt6-develop/third_party/mesagl/17.3.9/linux-x86_64_gcc-7.3/include' '-DOPENGL_gl_LIBRARY:STRING=/usr/workspace/justin/visit_builds/qt6-develop/third_party/mesagl/17.3.9/linux-x86_64_gcc-7.3/lib/libGL.so' '-DOPENGL_opengl_LIBRARY:STRING=' '-DOPENGL_glu_LIBRARY:STRING=/usr/workspace/justin/visit_builds/qt6-develop/third_party/mesagl/17.3.9/linux-x86_64_gcc-7.3/lib/libGLU.so' '-DOPENGL_GL_PREFERENCE:STRING=LEGACY' '-DBUILD_SHARED_LIBS=ON' '-DCMAKE_INSTALL_PREFIX=/usr/workspace/justin/visit_builds/qt6-develop/third_party/qt/6.4.2/linux-x86_64_gcc-7.3' '-DQT_BUILD_EXAMPLES=FALSE' '-DQT_BUILD_TESTS=FALSE' '-DCMAKE_BUILD_TYPE=Release' '-DINPUT_dbus=no' '-DINPUT_egl=no' '-DINPUT_eglfs=no' '-DINPUT_sql_db2=no' '-DINPUT_sql_ibase=no' '-DINPUT_sql_mysql=no' '-DINPUT_sql_oci=no' '-DINPUT_sql_odbc=no' '-DINPUT_sql_psql=no' '-DINPUT_sql_sqlite=no' '-DINPUT_libjpeg=no' '-DINPUT_libpng=qt' '-DINPUT_system_zlib=no' '-G' 'Unix Makefiles' '/usr/workspace/justin/visit_builds/qt6-develop/qtbase-everywhere-src-6.4.2'
-- Configuring incomplete, errors occurred!
CMake Error at /usr/workspace/justin/visit_builds/qt6-develop/qtbase-everywhere-src-6.4.2/cmake/QtProcessConfigureArgs.cmake:965 (message):
  CMake exited with code 1.


+ set +x
make: *** No targets specified and no makefile found.  Stop.
```
I guess it is mad about symlinks in the qt build dir? Strange. Did I set something up incorrectly?

biagas commented:
@JustinPrivitera 
Building qt5 came about from the --required.
I suppose using --qt6 should disable qt 5, good catch. I generally don't use the --required/--optional options, so I missed this scenario completely.


Re: qt6 build failure, what OS are you running on? I haven't seen that problem.
Any pertinent messages from the downloading of QT 6 ?

JustinPrivitera commented:
@biagas 
I am running on rztopaz which is still on toss3.

I see this from the download messages:
```
Checking for qt-5.14.2
Downloading qt-everywhere-src-5.14.2.tar.xz . . .
verifying SHA256 checksum c6fcd53c744df89e7d3223c02838a33309bd1c291fcb6f9341505fe99f7f19fa for qt-everywhere-src-5.14.2.tar.xz . . .
verified
Download succeeded: https://github.com/visit-dav/third-party/releases/download/v3.3.0/qt-everywhere-src-5.14.2.tar.xz
Checking for qwt-6.3.0
Downloading qwt-git-d3706f6e7f0351d278be2d989a4caaf92b399bbd.tar.xz . . .
verifying SHA256 checksum 39839f3aa83f41d09109296d41659e04bb234d9e41ab551af9f4e9b4fceed251 for qwt-git-d3706f6e7f0351d278be2d989a4caaf92b399bbd.tar.xz . . .
shasum -a 256 failed: looking for 39839f3aa83f41d09109296d41659e04bb234d9e41ab551af9f4e9b4fceed251 got 
Download attempt failed: https://github.com/visit-dav/third-party/releases/download/v3.3.0/qwt-git-d3706f6e7f0351d278be2d989a4caaf92b399bbd.tar.xz
verifying SHA256 checksum 39839f3aa83f41d09109296d41659e04bb234d9e41ab551af9f4e9b4fceed251 for qwt-git-d3706f6e7f0351d278be2d989a4caaf92b399bbd.tar.xz . . .
verified
```
Later we have this:
```
Checking for qt-6.4.2
Downloading qtbase-everywhere-src-6.4.2.tar.xz . . .
verifying SHA256 checksum a88bc6cedbb34878a49a622baa79cace78cfbad4f95fdbd3656ddb21c705525d for qtbase-everywhere-src-6.4.2.tar.xz . . .
shasum: qtbase-everywhere-src-6.4.2.tar.xz: 
shasum -a 256 failed: looking for a88bc6cedbb34878a49a622baa79cace78cfbad4f95fdbd3656ddb21c705525d got 
Download attempt failed: https://github.com/visit-dav/third-party/releases/download/v3.3.0/qtbase-everywhere-src-6.4.2.tar.xz
verifying SHA256 checksum a88bc6cedbb34878a49a622baa79cace78cfbad4f95fdbd3656ddb21c705525d for qtbase-everywhere-src-6.4.2.tar.xz . . .
verified
```

biagas commented:
@JustinPrivitera thanks for the info, I will investigate.

biagas commented:
@JustinPrivitera which compiler were you using? Qt6 requires gcc9/10, so I module loaded gcc/10.2.1 on rztopaz and didn't have any problems building qt6.
I will be pushing some changes soon that fix the problem with qt5 building  when --qt6 specified.

JustinPrivitera commented:
@biagas I was using gcc 7.3. Is there a way to make it quit and complain about the compiler version?

biagas committed:
Disable embree, ispc, tbb when building VTK 9.
Ospray 2.8 (used by vtk9) bundles these so separate builds aren't necessary.

biagas committed:
Disable pyside when building Qt 6.
Temporary measure until build module for pyside 6 can be implemented.

biagas committed:
Disable Qt(5) when building Qt 6.

biagas committed:
Merge branch 'develop' into task/biagas/add_qt6_support

biagas commented:
> @biagas I was using gcc 7.3. Is there a way to make it quit and complain about the compiler version?

Yes, bv_main.sh has a compiler version check. I can change the checks based on whether or not qt6 is being built, but I'm not sure what to do with different linux flavors (and versions of flavors) having different requirements (eg OpenSuse vs SuseEnterprise vs RH8 vs RH9 vs Ubuntu).
https://doc-snapshots.qt.io/qt6-6.5/supported-platforms.html#desktop-platforms

JustinPrivitera commented:
@biagas wow that looks like a real mess. Maybe the best thing to do is just to document the compiler requirements for the different systems and point users to that page you linked when they run into trouble.

I guess alternatively we could require the highest GCC version required on any system.. so I guess GCC11? Not sure if that is the best solution.

biagas commented:
@JustinPrivitera Yes, the QT supported platforms docs create quite a mess if I wanted to try comile-checking for them.  I'm not 100% certain that those levels of compilers are strictly necessary for building Qt from source. Might be the requirement for pre-built versions of Qt, which makes more sense. 

I don't think requiring the top g++ version (11) is a good idea, so I am going with the smallest I know will get qt to build, and adding information to the docs about possibly needing newer versions if the qt6 build fails.

FYI, I tried gcc-7.3 on rztopaz again, got a compiler error directly related to c++17 functionality, so 7.3 with qt6 is definitely a no-go.

biagas committed:
Add cfitsio 4.2.0 when building with MSVC2022.

biagas committed:
Merge branch 'develop' into task/biagas/add_qt6_support

biagas committed:
Update minimum compiler version checks for g++.

If building qt6, minimum g++ version is 8.1.

biagas committed:
Update build_visit docs for QT6.

Add links to Qt's Qt6 linux requirments.

Add section to Basic Usage, related to minimum compiler versions.

biagas committed:
Add comments regarding change from 'updateGL' to 'update' calls in ClipEditor files.

Qt 6 no longer has an 'updateGL' method, and the 'update' method seems to do the job.

biagas commented:
@JustinPrivitera I modified the minimum g++ version (if building with qt6)  in the compiler checks. This along with @markcmiller86's  fix to the version check in build_visit now has the script failing at the beginning if the compiler version is too old.  I  also added a brief statement regarding compiler versions to the Basic Usage page of the Building Visit docs. Let me know what you think.

biagas committed:
fix spacing in bullet list.

JustinPrivitera commented:
@biagas Sorry for the delay here.

I ran build visit with the same invocation as before:
```
./build_visit --makeflags -j24 --mesagl --mpich --qt6 --vtk --required --optional --no-gdal
```
qt6 fails to build:
```
Building QT6 base (~10 minutes)
Unzipping/Untarring qtbase-everywhere-src-6.4.2.tar.xz . . .
Configuring Qt6 base: . . . 
Building Qt6 base . . . 
Qt6 base build failed. Giving up
Unable to build or install QT6. Bailing out.
Error in build process. See build_visit_log for more information. If the error is unclear, please include build_visit_log and contact the VisIt project via https://visit-help.llnl.gov. You may need to compress the build_visit_log using a program like gzip so it will fit within the size limits for attachments.
Log file full path:  /usr/workspace/justin/visit_builds/qt6-develop/build_visit_log
```

I am using gcc 10.3.1 on rztopaz which is now on toss4.
```
bash-4.4$ gcc -v
Using built-in specs.
COLLECT_GCC=/usr/tce/packages/gcc/gcc-10.3.1/bin/gcc
...
```

These are the error messages from the qt6 download
```
Checking for qt-6.4.2
Downloading qtbase-everywhere-src-6.4.2.tar.xz . . .
verifying SHA256 checksum a88bc6cedbb34878a49a622baa79cace78cfbad4f95fdbd3656ddb21c705525d for qtbase-everywhere-src-6.4.2.tar.xz . . .
shasum: qtbase-everywhere-src-6.4.2.tar.xz: No such file or directory
shasum -a 256 failed: looking for a88bc6cedbb34878a49a622baa79cace78cfbad4f95fdbd3656ddb21c705525d got 
Download attempt failed: https://github.com/visit-dav/third-party/releases/download/v3.3.0/qtbase-everywhere-src-6.4.2.tar.xz
verifying SHA256 checksum a88bc6cedbb34878a49a622baa79cace78cfbad4f95fdbd3656ddb21c705525d for qtbase-everywhere-src-6.4.2.tar.xz . . .
verified
Download succeeded: https://download.qt.io/archive/qt/6.4/6.4.2/submodules/qtbase-everywhere-src-6.4.2.tar.xz
```

I'm seeing the same error as before when building qt6:
```
Building QT6 base (~10 minutes)
Build Qt 6 base module
prepare_build_dir: qtbase-everywhere-src-6.4.2 qtbase-everywhere-src-6.4.2.tar.xz
Unzipping/Untarring qtbase-everywhere-src-6.4.2.tar.xz . . .
Making build directory qtbase-everywhere-src-6.4.2-build
+ echo o
+ echo yes
+ tee qt.config.out
+ env PATH=/usr/workspace/justin/visit_builds/qt6-develop/third_party/cmake/3.24.3/linux-x86_64_gcc-10.3/bin:/usr/global/tools/jobutils/bin:/usr/tce/packages/mvapich2/mvapich2-2.3.7-intel-classic-2021.6.0-magic/bin:/usr/tce/packages/intel-classic/intel-classic-2021.6.0-magic/bin:/usr/tce/bin:/usr/lib64/ccache:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:. 'CFLAGS= -m64 -fPIC  -O2' 'CXXFLAGS= -m64 -fPIC  -O2' CC=gcc CXX=g++ ../qtbase-everywhere-src-6.4.2/configure -prefix /usr/workspace/justin/visit_builds/qt6-develop/third_party/qt/6.4.2/linux-x86_64_gcc-10.3 -no-dbus -no-egl -no-eglfs -no-sql-db2 -no-sql-ibase -no-sql-mysql -no-sql-oci -no-sql-odbc -no-sql-psql -no-sql-sqlite -no-libjpeg -qt-libpng -qt-zlib -nomake examples -nomake tests -opensource -confirm-license -shared -release -- -DOPENGL_INCLUDE_DIR:PATH=/usr/workspace/justin/visit_builds/qt6-develop/third_party/mesagl/17.3.9/linux-x86_64_gcc-10.3/include -DOPENGL_gl_LIBRARY:STRING=/usr/workspace/justin/visit_builds/qt6-develop/third_party/mesagl/17.3.9/linux-x86_64_gcc-10.3/lib/libGL.so -DOPENGL_opengl_LIBRARY:STRING= -DOPENGL_glu_LIBRARY:STRING=/usr/workspace/justin/visit_builds/qt6-develop/third_party/mesagl/17.3.9/linux-x86_64_gcc-10.3/lib/libGLU.so -DOPENGL_GL_PREFERENCE:STRING=LEGACY
Overriding option 'libjpeg' with 'no' (was: 'undefined')
opt: libpng val: qt
Overriding option 'libpng' with 'qt' (was: 'undefined')
opt: zlib val: qt
CMake Error at CMakeLists.txt:40 (message):
  The path
  "/usr/workspace/justin/visit_builds/qt6-develop/qtbase-everywhere-src-6.4.2-build"
  contains symlinks.  This is not supported.  Possible solutions:

              - map directories using a transparent mechanism such as mount --bind
              - pass the real path of the build directory to CMake, e.g. using             cd $(realpath <path>) before invoking cmake <source_dir>.
Call Stack (most recent call first):
  CMakeLists.txt:47 (qt_internal_check_if_path_has_symlinks)
```

What invocation did you use to build visit with qt6 on rz? Am I doing anything obviously wrong?

biagas commented:
@JustinPrivitera I am not sure what is going wrong, there shouldn't be any symlinks in `qtbase-everywhere-src-6.4.2-build`, as that subdirectory gets filled in by qmake/cmake.  Do you have environment variables related to qt? Or is there a Qt in your path that perhaps is causing problems?

Did you blow away the old `qtbase-everywhere-src-6.4.2-build` before you tried again?

To just test the building of Qt 6, I would invoke build_visit  with '--third-party --no-visit --cmake --qt6' 
I notice you didn't have `--cmake`,  is the build using a system cmake, then? Do you know which version?

JustinPrivitera commented:
@biagas

>  Do you have environment variables related to qt? Or is there a Qt in your path that perhaps is causing problems?

No and no.

> Did you blow away the old `qtbase-everywhere-src-6.4.2-build` before you tried again?

Yes; I started with a fresh directory.

> I notice you didn't have `--cmake`, is the build using a system cmake, then? Do you know which version?

cmake must be included under either `--required` or `--optional`, since it gets built as a result of my build invocation. It ends up building and using `cmake-3.24.3`.

biagas commented:
okay, what are the symlinks in that -build dir then? Perhaps knowing those would help to isolate the problem.

JustinPrivitera commented:
I am getting that same error each time with the symlinks, but there do not appear to be symlinks in that directory...

```
bash-4.4$ ls -lR qtbase-everywhere-src-6.4.2-build/
qtbase-everywhere-src-6.4.2-build/:
total 20
-rw------- 1 justin justin 4499 Jun  6 14:38 CMakeCache.txt
drwx--S--- 3 justin justin 4096 Jun  6 14:38 CMakeFiles
-rw------- 1 justin justin  327 Jun  6 14:38 config.opt
-rw------- 1 justin justin  741 Jun  6 14:38 qt.config.out

qtbase-everywhere-src-6.4.2-build/CMakeFiles:
total 8
-rw------- 1 justin justin   85 Jun  6 14:38 cmake.check_cache
drwx--S--- 2 justin justin 4096 Jun  6 14:38 pkgRedirects

qtbase-everywhere-src-6.4.2-build/CMakeFiles/pkgRedirects:
total 0
bash-4.4$ ls -lR qtbase-everywhere-src-6.4.2-build/ | grep ^l
bash-4.4$ 
```

Odd.

biagas commented:
Well, the only thing I can think of now, is that '/usr/workspace/justin' is a symlink to '/usr/WS2/justin'.
Somehow the build system is using the symlink instead of the resolved path.
Not sure why that doesn't happen to me when I build on rztopaz. Gonna try an experiment.

biagas commented:
> Well, the only thing I can think of now, is that '/usr/workspace/justin' is a symlink to '/usr/WS2/justin'. Somehow the build system is using the symlink instead of the resolved path. Not sure why that doesn't happen to me when I build on rztopaz. Gonna try an experiment.

@JustinPrivitera  Yes, the problem is `/usr/workspace/justin`.  I tried this:
```cd  /usr/workspace/kbonnell/mybuild```
and ran the same build_visit invocation as you, and I get the symlink error,
if I instead
```cd  /usr/WS1/kbonnell/mybuild``` and run build_visit, I don't get the symlink error.

I usually use the full path when I navigate the filesystem and not the symlink, so that's why I didn't run across this before.

I will check bv_qt6 and see if I can find a way to mitigate this problem.

JustinPrivitera commented:
Wow that is sneaky. I don't think I would have thought of that.

biagas committed:
Fix qt6 build failure when symlinks present in build dir.

biagas commented:
@JustinPrivitera   Qt actually gave us a clue:

``` pass the real path of the build directory to CMake, e.g. using             cd $(realpath <path>)``` 

I modified bv_qt6.sh to use `cd -P ${QT6_BASE_BUILD_DIR}` instead of just `cd ${QT6_BASE_BUILD_DIR}`. 
I was concerned that 'realpath' might not be available everywhere.
At any rate, the change resolved the issue for me, can you give it a whirl?

JustinPrivitera commented:
Now it succeeds in building all TPLs, but fails to build visit proper:

```
CMake Error at /usr/lib64/cmake/Qt5/Qt5Config.cmake:28 (find_package):
  Could not find a package configuration file provided by "Qt5UiTools" with
  any of the following names:

    Qt5UiToolsConfig.cmake
    qt5uitools-config.cmake

  Add the installation prefix of "Qt5UiTools" to CMAKE_PREFIX_PATH or set
  "Qt5UiTools_DIR" to a directory containing one of the above files.  If
  "Qt5UiTools" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMake/FindVisItQt.cmake:85 (find_package)
  CMake/SetUpThirdParty.cmake:565 (include)
  CMakeLists.txt:919 (include)


-- Configuring incomplete, errors occurred!
See also "/usr/workspace/justin/visit_builds/qt6-develop/visit/build/CMakeFiles/CMakeOutput.log".
See also "/usr/workspace/justin/visit_builds/qt6-develop/visit/build/CMakeFiles/CMakeError.log".
```

I wonder why the qt5 stuff is even run?

biagas commented:
Sorry about that. I probably didn't have build_visit set up the host profile correctly to use Qt6 and not Qt5.
Will look into that. Again, thanks for testing the build module!

biagas committed:
Fix host profile issues when using --qt6.

biagas committed:
Merge branch 'develop' into task/biagas/add_qt6_support

biagas commented:
@JustinPrivitera I fixed issues with the host-profile creation: qt5 was added to host profile even though --qt6 was requested, and --qt6 wasn't adding the 'SETUP_APPP_VERSION' call which tells the find module which version of Qt to look for.

Building of VisIt will still fail using the build_visit script, because it will check out a copy of develop, but you need my branch in order to have the QT6 stuff in it.  If you still want to check the build, grab my changes to build_visit, and re-run. Let the build of visit fail, then cd to the visit dir and checkout branch associated with this PR: task/biagas/add_qt6_suppport, Then run build_visit again.

biagas committed:
Fix typo in Qt 6 version var.

JustinPrivitera commented:
I see this inside the config site file:
```
##
## QT6
##
SETUP_APP_VERSION(QT 5.14.2)
VISIT_OPTION_DEFAULT(VISIT_QT_DIR ${VISITHOME}/qt/6.4.2/${VISITARCH})
```

This seems at odds with your latest changes; could I be doing something wrong?


Also, I am using 
```
./visit/src/tools/dev/scripts/build_visit --write-unified-file build_visit
```
to make my build visit script from your branch, so my understanding is that when I run it, it should be building your branch. Is that right?

biagas commented:
Just fixed that! (the host profile qt version problem).

build_visit  from my branch will build all TP libraries according to the changes from the branch, but visit itself will be a checkout of develop.  I don't think we currently have a way to tell build_visit to checkout a particular branch.

JustinPrivitera commented:
Ok I am running now; it seems to be getting farther than before. Made it past configuration.

> I don't think we currently have a way to tell build_visit to checkout a particular branch.

What I've been doing is cloning the visit repo, changing to my favorite branch, and then creating build visit from that branch and running. I don't think build visit changes the branch back to develop, so I think in that way I've been able to build my branches with TPLs with little hassle. That's just doing what you suggested but in a different order, right?

biagas commented:
> What I've been doing is cloning the visit repo, changing to my favorite branch, and then creating build visit from that branch and running. I don't think build visit changes the branch back to develop, so I think in that way I've been able to build my branches with TPLs with little hassle. That's just doing what you suggested but in a different order, right?

Sounds right.  Definitely different than my workflow, but makes more sense, especially if you want build_visit to build visit itself (which I don't often do).

JustinPrivitera commented:
Ok, so it builds successfully, but the main window is broken:

![image](https://github.com/visit-dav/visit/assets/35237779/4c6cd13a-1de8-4193-8def-a2ff9867bc75)

No matter what I do, it stays black.

biagas commented:
> No matter what I do, it stays black.
Yep, that's a VTK-9 with Mesa issue, and Qt 6 requires VTK 9. That issue is still being worked on.

JustinPrivitera commented:
So in that case, if I can build successfully, then everything is probably good?

> That issue is still being worked on.

For now, if a user runs build visit with `--qt6` will this happen no matter what?

biagas commented:
I think things are fine with a good enough system GL, but if you have to use Mesa, then that's when things go awry.
If you use qt5 but ask for --vtk9 and --mesagl you will have the same problem. (So it wasn't introduced by this PR).

Since this is on develop, and the kinks should be worked out before we put it in a release, I'm not sure if any mitigating steps should be taken, like a warning in build_visit if --mesagl and ( --vtk9 or --qt6) are requested. What do you think?

JustinPrivitera commented:
If I run build visit with `--osmesa` instead it should work properly?

biagas commented:
I don't think so.

JustinPrivitera commented:
Sorry. Is there a combination of build arguments that will give me visit with qt6 and vtk9 that will not have the black screen?

biagas commented:
> Sorry. Is there a combination of build arguments that will give me visit with qt6 and vtk9 that will not have the black screen?

On a system with sufficient graphics drivers and opengl dev (3.2)?, (like my windows machine, or a desktop linux)  just use the same args without --mesagl.

biagas commented:
@JustinPrivitera Thank you very much for you diligence in this review.

JustinPrivitera commented:
@biagas glad I could assist :)

JustinPrivitera commented:
When I built without `--mesagl` I get a visit that looks like this:
![image](https://github.com/visit-dav/visit/assets/35237779/b7916349-7187-4db9-87c8-0114a468a4fe)

Not wrong, just different.

biagas commented:
@JustinPrivitera thanks for the screenshot.  Was this a full rebuild of TP libs, or just Qt , or just VisIt?

JustinPrivitera commented:
@biagas started from scratch: full rebuild of TPLs and visit with this build visit invocation:
`./build_visit --makeflags -j24 --mpich --qt6 --vtk --required --optional --no-gdal`

JustinPrivitera commented:
@biagas  do you think this problem warrants me opening a ticket about it? It seems pretty minor...

biagas commented:
> @biagas do you think this problem warrants me opening a ticket about it? It seems pretty minor...

It is probably a good idea to keep track of issues like this.   In your notes include the system, your build_visit args, and just flag the fact that it is VTK9 and Qt6 combined. Don't know which library is the source of the problem.
